File: android/app/src/main/java/com/darkraha/gldemos/renders/TexturedQuadRender.kt
package com.darkraha.gldemos.renders

import android.content.Context
import android.opengl.GLES30.*
import com.darkraha.gldemos.R
import com.darkraha.gldemos.gl.*
import org.joml.Matrix4f
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

/**
 * Render textured quad. Demonstrate how to apply 2D texture to the surface.
 *
 */
class TexturedQuadRender(val context: Context) : Render() {

    private lateinit var prog: ShaderProgram
    private lateinit var texture: GlTexture
    private lateinit var matrix: Matrix4f

    private var idVao = 0
    private var idVbo = 0

    private val rotY = 1.5f * GlCommon.TO_RAD
    private val rotX = GlCommon.TO_RAD

    override fun onSurfaceCreated(gl: GL10?, config: EGLConfig?) {
        glClearColor(0.0f, 0.0f, 0.0f, 0.0f)
        glEnable(GL_DEPTH_TEST)

        prog = ShaderProgramBuilder()
            .vertexAttributes(false, true, false)
            .build()

        matrix = Matrix4f()
            .perspective(GlCommon.ALNGLE45, aspect, 1f, 100f)
            .translate(0f, 0f, -6f)

        glUseProgram(prog.idProgram)
        texture = GlTexture.newTexture2D(R.drawable.t235, "test")
        idVao = GlUtils.createVAO()
        idVbo = GlUtils.createVBO(
            floatArrayOf( // coords         texture coords
                -1.0f, -1.0f,  /**/0.0f, 1.0f,
                1.0f, -1.0f,  /**/1.0f, 1.0f,
                1.0f, 1.0f,  /**/1.0f, 0.0f,
                1.0f, 1.0f,  /**/1.0f, 0.0f,
                -1.0f, 1.0f,  /**/0.0f, 0.0f,
                -1.0f, -1.0f,  /**/0.0f, 1.0f
            )
        )

        GlUtils.bindAttributes(idVbo, 2, 0, true)
    }


    override fun onSurfaceChanged(gl: GL10?, width: Int, height: Int) {
        super.onSurfaceChanged(gl, width, height)
        matrix = Matrix4f()
            .perspective(GlCommon.ALNGLE45, aspect, 1f, 100f)
            .translate(0f, 0f, -8f)
    }

    override fun onDrawFrame(arg0: GL10?) {
        glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)

        glUseProgram(prog.idProgram)

        //-----------------------------------------------
        // activate our texture
        // you can use replace it by prog.uniformTexture(texture);
        val samplerLocation = glGetUniformLocation(prog.idProgram, ShaderProgramBuilder.U_SAMPLER_NAME)
        glActiveTexture(GL_TEXTURE0)
        glBindTexture(GL_TEXTURE_2D, texture.idTexture)
        glUniform1i(samplerLocation, 0)

        matrix.rotateAffineXYZ(rotX, rotY, 0f)
        prog.uniformMatrix(matrix)

        glBindVertexArray(idVao)
        // glEnable(GL_CULL_FACE);
        glDrawArrays(GL_TRIANGLES, 0, 6) // vertex count
    }

    override fun onDispose() {
        glUseProgram(0)
        prog.dispose()
        texture.dispose()
        glDeleteVertexArrays(1, intArrayOf(idVao),0)
        glDeleteBuffers(1, intArrayOf(idVbo),0)
    }
}

File: webgl/source/js/renders/texturedquadrender.js
class TexturedQuadRender extends Render {


    matrix
    prog
    idVao
    idVbo
    rotY = 1.5 * Math.PI / 180.0;
    rotX = Math.PI / 180.0;
    texture


    onSetup(appOGL) {
        super.onSetup(appOGL);
        const gl = appOGL.gl;

        const prog = new ShaderProgramBuilder(gl)
            .texture2D()
            .build();
        this.prog = prog;

        this.matrix = glMatrix.mat4.create();
        glMatrix.mat4.perspective(this.matrix, 45 * Math.PI / 180.0,
            gl.canvas.clientWidth / gl.canvas.clientHeight,
            0.1, 100.0
        );

        glMatrix.mat4.translate(this.matrix, this.matrix, [-0.0, 0.0, -6.0]);

        gl.useProgram(prog.idProgram);

        this.texture = GlTexture.newTexture2D(gl, "images/textures/235.jpg", "test");

        const vData = [
            // coords         texture coords
            -1.0, -1.0, /**/ 0.0, 1.0,
            1.0, -1.0, /**/ 1.0, 1.0,
            1.0, 1.0, /**/ 1.0, 0.0,
            1.0, 1.0, /**/ 1.0, 0.0, -1.0, 1.0, /**/ 0.0, 0.0, -1.0, -1.0, /**/ 0.0, 1.0
        ];

        this.idVao = GlUtils.createVAO(gl);
        this.idVbo = GlUtils.createVBO(gl, vData);
        GlUtils.bindAttributes(gl, this.idVbo, 2, 0, true);
    }

    onDrawFrame(appOGL) {
        super.onDrawFrame(appOGL);
        const mat4 = glMatrix.mat4;
        const matrix = this.matrix;
        const gl = appOGL.gl;
        const prog = this.prog;
        const TO_RAD = Math.PI / 180.0;


        mat4.rotate(matrix, matrix, this.rotX, [1, 0, 0]);
        mat4.rotate(matrix, matrix, this.rotY, [0, 1, 0]);

        gl.useProgram(prog.idProgram);
        gl.uniformMatrix4fv(gl.getUniformLocation(prog.idProgram, 'm'), false, matrix);


        // Tell WebGL we want to affect texture unit 0
        gl.activeTexture(gl.TEXTURE0);

        // Bind the texture to texture unit 0
        gl.bindTexture(gl.TEXTURE_2D, this.texture.idTexture);

        const sampleLocation = gl.getUniformLocation(prog.idProgram, 'sampler');

        // Tell the shader we bound the texture to texture unit 0
        gl.uniform1i(sampleLocation, 0);

        gl.bindVertexArray(this.idVao);

        // gl.enable(gl.CULL_FACE);
        gl.drawArrays(gl.TRIANGLES, 0, 6);
    }

}